compiler/plugins/target/AMD-AIE/iree-amd-aie/IR/AMDAIETraits.h

```diff
@@ -0,0 +1,22 @@
+// Copyright 2024 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef IREE_COMPILER_AMDAIE_TRAITS_H_
+#define IREE_COMPILER_AMDAIE_TRAITS_H_
+
+#include "mlir/IR/OpDefinition.h"
+
+namespace mlir::OpTrait::iree_compiler::AMDAIE {
+
+template <typename ConcreteType>
+class CircularDmaOp : public OpTrait::TraitBase<ConcreteType, CircularDmaOp> {
```

Comment:
Just curious: why a trait rather than an attribute?

Reply:
Isn't this what `Traits` are meant for, to describe common properties of operations? I am trying to do something similar like `MemoryEffects`, but for our DMA operations. And ideally, I would like to find a good way to annotate the operands (with attributes?) as the resources the trait relates to, for example:

```
MyOp [CircularDmaOp] {
  ins Arg<Index, "", DmaRead>:$target,
       Arg<Index, "", DmaWrite>:$source
}
```

Then, it would be possible to create more general CSE logic to check whether another DMA uses the same resources and in what way.

Something like this seems to exist already for memory effects, but I don't how to use it, I don't see any code getting generated for it either when I tried using it and I don't see CSE leveraging it either..

Reply:
> Isn't this what Traits are meant for, to describe common properties of operations?

It's functionality rather than properties though - like it's cpp level stuff rather then IR/compiler/domain level stuff. Now yes, naturally sometimes those things go hand-in-hand e.g. TilingInterface/Traits naturally apply only to a certain kind of op (loopy things) but the Trait itself is for giving connections/hooks for how something happens in cpp land - how the IR is rewritten by cpp methods.

> similar to MemoryEffects

yea so I'm guessing the way this works is somewhere in the CSE system there's a check for like RAW/WAR/WAW or whatever so it's cpp level but I think it's a poor use, specifically because

> And ideally, I would like to find a good way to annotate the operands

traits aren't visible in the IR itself (they're visible in ODS)

> Something like this seems to exist already for memory effects, but I don't how to use it, I don't see any code getting generated for it when I tried using it and I don't see CSE leveraging it either..

I'll do some digging some time tomorrow

Reply:
btw as always this isn't biblical scripture/dogma - just my intuition

Reply:
> traits aren't visible in the IR itself (they're visible in ODS)

That's fine, I don't think I need it to be